### DISCUSSION TITLE
some thoughts about rewrite/communicate...

### DISCUSSION THREAD

**Discussion by silas1037** [Stellarium/stellarium#1426] [General]
The greate stellarium was written by cpp, and I also found some earlier rewritten version —— like Stellarium for Java.

The following are just some ideas...

1. Is that possible/easy to rewrite / port it  to other language like python? 

Rewrite: Translating a OOJ to OOJ seems to be easier more, but simpler language's code will be shorter... 

2. Although the cpp-base program success at running speed, to share it calculation data or communication between different programs also values.

Functions to communicate with diff language by interface: handle a call from other programs, returning what they wants like astro calculation results. Well, remote-control can work for it.

---

On some website I also saw the headless mode of none-gui stellarium in docker, maybe port code to a dynamic lib is possible..

**Comment by gzotti** [comment]
Please see the permanent success of Stellarium for Java for an answer.

**Comment by silas1037** [comment]
> Please see the permanent success of Stellarium for Java for an answer.

I don't understand the words. S4J has sinked for so long.

**Comment by alex-w** [comment]
Short answer: you should start new project for resolving your ideas (some dynamic library and GUI as a plugin or standalone application for that library), becasue this will be simpler that rewritting 80% of planetarium to obtaining dynamic library with some API.

P.S. I meant a classic dynamic library

**Comment by alex-w** [comment]
One of possible solution for adding ability to an interoperability of planetarium is adding some interface, like DBUS. But (!) this solution will not switch Stellarium to mode of work, like some guiless library for calculations.

**Reply by silas1037** [reply]
I guess skipping the opengl render is feasible.

**Reply by gzotti** [reply]
The way to interact with the current system, demonstrated successfully in several external applications, is via RemoteControl API. I assume DBUS communication would work in a similar fashion by interfacing with module properties, just that somebody will have to develop it. With WSL's success this may even become interesting on the Windows OS. It may be possible to interface with Python with the approach from MevisLab. These are a lot of "may"s however. 

A full rewrite in a "simpler" language (is Python really simpler?) depends on somebody doing it and then maintaining it, hence my satiric suggestion of looking at S4J's permanent success.

**Reply by gzotti** [reply]
Skipping the OpenGL renderer without a replacement (DirectX, Vulkan, whatever...) is just taking away the very best part. There are many general astro commandline apps or web tools.

**Reply by silas1037** [reply]
I agree.
